File: tests/js/test_firefox_auto_panel.mjs
/**
 * The 18th window `firefox_auto` — "Firefox auto with Extension" (I-63).
 *
 * Part 1: the mount — `winFirefoxAuto` is registered in every registry
 * (Python catalog, constants.js, store.js winElIds, _PANEL_INITS) and mounts
 * inside the grid, never an L-5 orphan. Part 2: the content — the config
 * fields (the pattern is the control element), Save/Run/Stop over the 4 slots
 * and the live `firefox_auto_updated` step stream ending in one named verdict.
 */
import { test, describe } from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import vm from 'node:vm';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { El } from './fake_dom.mjs';
import { createSashGrid, ALL_WINDOW_IDS, panelIdOf } from './sash_harness.mjs';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const WEB = path.resolve(__dirname, '../../app/ui/web');
const readJs = (rel) => fs.readFileSync(path.join(WEB, 'js', rel), 'utf-8');
const html = fs.readFileSync(path.join(WEB, 'index.html'), 'utf8');

const FA_IDS = ['faPattern', 'faUrl', 'faMacro', 'faTarget', 'faStorage', 'faHome', 'faBinary', 'faMode',
  'faTimeout', 'faPause', 'faSaveBtn', 'faRunBtn', 'faStopBtn', 'faState', 'faPaths',
  'faStatus', 'faSteps'];

function inside(node, ancestor) {
  for (let n = node; n; n = n.parent) if (n === ancestor) return true;
  return false;
}

describe('firefox_auto window (mount)', () => {
  test('index.html registers winFirefoxAuto with the config fields and controls', () => {
    assert.ok(html.includes('id="winFirefoxAuto" data-window="firefox_auto"'), 'panel present');
    assert.ok(html.includes('js/panels/firefox-auto.js'), 'index.html loads the panel script');
    const start = html.indexOf('id="winFirefoxAuto"');
    const end = html.indexOf('<div class="panel', start + 1);
    const block = html.slice(start, end > 0 ? end : undefined);
    for (const id of FA_IDS) assert.ok(block.includes(`id="${id}"`), `${id} inside winFirefoxAuto`);
    assert.match(block, /Firefox auto with Extension/);
    assert.match(block, /faPattern/);   // the user-set pattern is the control element
  });

  test('the grid mounts winFirefoxAuto and does not destroy it on render()', () => {
    assert.ok(ALL_WINDOW_IDS.includes('firefox_auto'), 'harness registry has the 18th window');
    assert.equal(panelIdOf('firefox_auto'), 'winFirefoxAuto');
    const h = createSashGrid();
    const find = (n, id) => { if (n.id === id) return n; for (const c of n.children) { const r = find(c, id); if (r) return r; } return null; };
    h.SashGrid.render();
    const win = find(h.body, 'winFirefoxAuto');
    assert.ok(win, 'winFirefoxAuto exists after render');
    assert.ok(inside(win, h.gridEl), 'winFirefoxAuto is inside #sashGrid (not discarded by replaceChildren)');
  });

  test('_PANEL_INITS includes FirefoxAutoPanel and it publishes itself on window', () => {
    const inits = readJs('arena-app.js').match(/_PANEL_INITS\s*=\s*\[([\s\S]*?)\]/)[1];
    assert.match(inits, /'FirefoxAutoPanel'/);
    const sandbox = { console: { warn() {}, log() {}, error() {} }, JSON, Object, Array, Set, WeakMap, Map, Number, String, Math, Date, parseInt, isNaN,
      setInterval: () => 1, setTimeout: () => 1, document: { getElementById: () => null, readyState: 'complete', addEventListener() {} } };
    sandbox.window = sandbox;
    vm.createContext(sandbox);
    vm.runInContext(readJs('core/boot.js'), sandbox, { filename: 'core/boot.js' });
    vm.runInContext(readJs('panels/firefox-auto.js'), sandbox, { filename: 'panels/firefox-auto.js' });
    assert.ok(vm.runInContext("Boot.panel('FirefoxAutoPanel')", sandbox), 'published on window');
    vm.runInContext('FirefoxAutoPanel.init()', sandbox); // no DOM, no bridge: must not throw
  });
});

/* ── part 2: the content ─────────────────────────────────────────────────── */
const CFG = { pattern: 'Arena', url: 'https://arena.ai', target: "xpath=//a[span[text()='New Chat']]",
  macro: 'Python_XClick_Demo', storage: 'xfile', home: '', binary: '', timeout_sec: 90, pause_ms: 3000, mode: 'find' };
const PATHS = { home: '/home/u/Desktop/uivision',
  macro_file: '/home/u/Desktop/uivision/macros/Python_XClick_Demo.json',
  autorun_file: '/cfg/uivision/ui.vision.html', log_dir: '/cfg/uivision/logs' };

function bootFaPanel() {
  const byId = {};
  for (const id of FA_IDS) byId[id] = new El(id.endsWith('Btn') ? 'button' : 'div');
  const calls = [];
  const listeners = {};
  const responses = {
    get_firefox_auto_config: JSON.stringify({ ok: true, config: CFG, paths: PATHS, running: false }),
    save_firefox_auto_config: JSON.stringify({ ok: true, config: CFG, paths: PATHS }),
    run_firefox_auto_test: JSON.stringify({ ok: true, state: 'running' }),
    stop_firefox_auto_test: JSON.stringify({ ok: true, state: 'stopping' }),
  };
  const bridge = new Proxy({}, { get(_t, k) {
    if (typeof k !== 'string') return undefined;
    const fn = (...a) => { calls.push({ name: k, args: a }); const cb = a.find((x) => typeof x === 'function'); if (cb) cb(responses[k] || '{}'); };
    fn.connect = (h) => { (listeners[k] ||= []).push(h); };
    return fn;
  } });
  const sandbox = {
    console, JSON, Object, Array, Math, Number, String, Set, Map, WeakMap, parseInt, parseFloat, isNaN,
    setTimeout: () => 1, clearTimeout() {}, setInterval: () => 1, clearInterval() {},
    document: { getElementById: (id) => byId[id] || null, querySelector: () => null, querySelectorAll: () => [],
      readyState: 'complete', addEventListener() {}, createElement: (t) => new El(t) },
    App: { bridge },
  };
  sandbox.window = sandbox;
  vm.createContext(sandbox);
  vm.runInContext(readJs('core/boot.js'), sandbox, { filename: 'core/boot.js' });
  vm.runInContext(readJs('panels/firefox-auto.js'), sandbox, { filename: 'panels/firefox-auto.js' });
  vm.runInContext('FirefoxAutoPanel.init()', sandbox);
  const push = (payload) => { for (const h of listeners.firefox_auto_updated || []) h(JSON.stringify(payload)); };
  const call = (name) => calls.find((c) => c.name === name);
  const click = (id) => byId[id].dispatch('click');
  return { byId, calls, responses, push, call, click, sandbox,
    text: (id) => byId[id].textContent };
}

describe('firefox_auto window (content)', () => {
  test('the first paint pulls the config and fills every field, the paths and idle state', () => {
    const h = bootFaPanel();
    assert.ok(h.call('get_firefox_auto_config'), 'initial pull on bridge-ready');
    assert.equal(h.byId.faPattern.value, 'Arena');
    assert.equal(h.byId.faTarget.value, CFG.target);
    assert.equal(h.byId.faTimeout.value, 90);
    assert.equal(h.byId.faPause.value, 3000);
    assert.match(h.text('faPaths'), /Ui\.Vision home: \/home\/u\/Desktop\/uivision/);
    assert.match(h.text('faPaths'), /macro file: .*Python_XClick_Demo\.json/);
    assert.equal(h.text('faState'), 'idle');
    assert.ok(!h.byId.faRunBtn.disabled);
  });

  test('Save sends exactly the field table (numbers as numbers) and shows the verdict', () => {
    const h = bootFaPanel();
    h.byId.faPattern.value = 'My Pattern';
    h.click('faSaveBtn');
    const save = h.call('save_firefox_auto_config');
    assert.ok(save, 'save slot called');
    const payload = JSON.parse(save.args[0]);
    assert.deepEqual(Object.keys(payload).sort(), Object.keys(CFG).sort());
    assert.equal(payload.pattern, 'My Pattern');
    assert.equal(payload.timeout_sec, 90);
    assert.equal(typeof payload.timeout_sec, 'number');
    assert.equal(h.text('faStatus'), '✅ config saved');
    assert.equal(h.byId.faStatus.style.color, '#4caf50');
  });

  test('a refused save names the error and stays visible', () => {
    const h = bootFaPanel();
    h.responses.save_firefox_auto_config = JSON.stringify({ ok: false, error: 'macro name not allowed' });
    h.click('faSaveBtn');
    assert.match(h.text('faStatus'), /⚠ save refused: macro name not allowed/);
    assert.equal(h.byId.faStatus.style.color, '#e6a23c');
  });

  test('Run clears the steps, disables itself and streams every phase', () => {
    const h = bootFaPanel();
    h.push({ kind: 'step', step: 'provision', message: 'macro written: /h/macros/M.json' });
    h.click('faRunBtn');
    assert.ok(h.call('run_firefox_auto_test'), 'run slot called');
    assert.equal(h.text('faSteps'), '', 'old steps cleared on a new run');
    assert.equal(h.text('faState'), 'running…');
    assert.equal(h.byId.faRunBtn.disabled, true);
    h.push({ kind: 'step', step: 'launch', message: 'starting Firefox', level: 'info' });
    assert.match(h.text('faSteps'), /· launch: starting Firefox/);
    assert.equal(h.text('faStatus'), 'launch — starting Firefox');
    h.push({ kind: 'step', step: 'foreground', message: 'no Firefox window matches', level: 'warn' });
    h.push({ kind: 'step', step: 'launch', message: 'Firefox not found', level: 'error' });
    assert.match(h.text('faSteps'), /⚠ foreground: no Firefox window matches/);
    assert.match(h.text('faSteps'), /✗ launch: Firefox not found/);
  });

  test('the ok result re-enables Run and appends the macro log tail', () => {
    const h = bootFaPanel();
    h.click('faRunBtn');
    h.push({ kind: 'result', result: 'ok', message: 'macro completed', lines: ['echo: done — XClick fired (native OS input)'] });
    assert.equal(h.text('faState'), 'idle');
    assert.ok(!h.byId.faRunBtn.disabled);
    assert.match(h.text('faStatus'), /✅ ok: macro completed/);
    assert.match(h.text('faSteps'), /\| echo: done — XClick fired/);
  });

  test('every verdict kind shows its own icon — never one invented "failed"', () => {
    for (const [result, icon] of [['timeout', '⌛'], ['stopped', '⏹'], ['blocked', '❌'], ['error', '❌']]) {
      const h = bootFaPanel();
      h.click('faRunBtn');
      h.push({ kind: 'result', result, message: 'because' });
      assert.match(h.text('faStatus'), new RegExp(`${icon} ${result}: because`), result);
      assert.equal(h.text('faState'), 'idle', `${result} ends the run`);
    }
  });

  test('Stop asks the backend and says what will happen', () => {
    const h = bootFaPanel();
    h.click('faStopBtn');
    assert.ok(h.call('stop_firefox_auto_test'), 'stop slot called');
    assert.match(h.text('faStatus'), /⏹ stop requested — the run ends on its next check/);
  });

  test('a pushed running/saved payload repaints without any bridge call', () => {
    const h = bootFaPanel();
    const before = h.calls.length;
    h.push({ kind: 'running' });
    assert.equal(h.text('faState'), 'running…');
    h.push({ kind: 'saved', config: { pattern: 'Z' }, paths: { home: '/h' } });
    assert.equal(h.byId.faPattern.value, 'Z');
    assert.match(h.text('faPaths'), /Ui\.Vision home: \/h/);
    assert.equal(h.calls.length, before, 'pushed payloads never call the bridge');
  });

  test('the step stream is capped at the last 60 lines', () => {
    const h = bootFaPanel();
    for (let n = 1; n <= 65; n++) h.push({ kind: 'step', step: 's', message: `m${n}` });
    const lines = h.text('faSteps').split('\n');
    assert.equal(lines.length, 60);
    assert.match(lines[59], /m65/);
    assert.match(lines[0], /m6/);
  });
});

File: tests/test_firefox_auto_panel.py
"""Firefox-auto panel — config get/save, run/stop slots, the status stream.

Stub-bridge tests (the test_job_history.py pattern): real ConfigManager on
tmp_path, FakeSignal records every `firefox_auto_updated` payload, and the run
slot is exercised through its schedule_coro + runner seams — no Qt, no Firefox.
"""

import json
from types import SimpleNamespace

import pytest

from app.browser.uivision import runner as uiv_runner
from app.persistence.config_manager import DEFAULT_SESSION, ConfigManager
from app.ui.panels import firefox_auto as fa
from app.ui.panels.firefox_auto import FirefoxAutoMixin

pytestmark = pytest.mark.unit

DEFAULTS = DEFAULT_SESSION["firefox_auto"]


class FakeSignal:
    def __init__(self):
        self.sent = []

    def emit(self, payload):
        self.sent.append(payload)

    def connect(self, _fn):
        pass


def make_bridge(tmp_path, **kw):
    logs = []
    ns = {"_log": lambda m, l="info": logs.append((l, m)), "_logs": logs,
          "firefox_auto_updated": FakeSignal(),
          "config": ConfigManager(str(tmp_path / "cfg"))}
    ns.update(kw)
    return SimpleNamespace(**ns)


def payloads(bridge):
    return [json.loads(p) for p in bridge.firefox_auto_updated.sent]


# ── get/save config ──────────────────────────────────────────────────────────

def test_get_config_returns_defaults_paths_and_idle_flag(tmp_path):
    fake = make_bridge(tmp_path)
    res = json.loads(FirefoxAutoMixin.get_firefox_auto_config(fake))
    assert res["ok"] is True and res["running"] is False
    assert res["config"] == DEFAULTS
    paths = res["paths"]
    assert set(paths) == {"home", "macro_file", "autorun_file", "log_dir"}
    assert paths["macro_file"].endswith("macros/Python_XClick_Demo.json")
    assert paths["autorun_file"].endswith("ui.vision.html")


def test_save_validates_clamps_and_persists(tmp_path):
    fake = make_bridge(tmp_path)
    res = json.loads(FirefoxAutoMixin.save_firefox_auto_config(fake, json.dumps({
        "pattern": "  My Pattern ", "url": "https://example.com", "target": "xpath=//a",
        "macro": "Ok-Macro_2", "storage": "WEIRD", "home": "", "binary": "",
        "timeout_sec": 9999, "pause_ms": "junk"})))
    assert res["ok"] is True
    cfg = res["config"]
    assert cfg["pattern"] == "My Pattern"                 # trimmed
    assert cfg["macro"] == "Ok-Macro_2"
    assert cfg["storage"] == "xfile"                      # unknown → default
    assert cfg["timeout_sec"] == 600                      # clamped to TIMEOUT_RANGE
    assert cfg["pause_ms"] == DEFAULTS["pause_ms"]        # garbage → default
    assert "url" not in cfg                               # retired key never comes back
    assert fake.config.get_state("firefox_auto") == cfg   # persisted
    saved = payloads(fake)[-1]
    assert saved["kind"] == "saved" and saved["config"] == cfg and "paths" in saved
    assert any(lvl == "success" and "config saved" in msg for lvl, msg in fake._logs)


def test_retired_url_key_drops_from_an_old_stored_config(tmp_path):
    """Old session.json files still carry the pre-2026-09-23 "url" field."""
    fake = make_bridge(tmp_path)
    fake.config.set_state(**{"firefox_auto": dict(DEFAULTS, url="https://arena.ai")})
    cfg = fa.load_config(fake)
    assert "url" not in cfg                               # load heals (RULE 13)
    assert "url" not in fa.validate_config({"url": "https://arena.ai"})   # RULE 10
    res = json.loads(FirefoxAutoMixin.save_firefox_auto_config(fake, "{}"))
    assert res["ok"] is True and "url" not in res["config"]
    assert "url" not in fake.config.get_state("firefox_auto")


def test_save_refuses_a_bad_macro_name_by_name(tmp_path):
    fake = make_bridge(tmp_path)
    res = json.loads(FirefoxAutoMixin.save_firefox_auto_config(fake, json.dumps(
        {"macro": "bad name"})))
    assert res["ok"] is False and "not allowed" in res["error"]
    assert fake.config.get_state("firefox_auto") == DEFAULTS     # nothing persisted
    assert any(lvl == "warn" and "refused" in msg for lvl, msg in fake._logs)


def test_save_refuses_unreadable_json(tmp_path):
    fake = make_bridge(tmp_path)
    res = json.loads(FirefoxAutoMixin.save_firefox_auto_config(fake, "{oops"))
    assert res["ok"] is False and "unreadable JSON" in res["error"]


def test_load_heals_a_stored_bad_macro_and_unknown_keys(tmp_path):
    fake = make_bridge(tmp_path)
    fake.config.set_state(**{"firefox_auto": {"macro": "bad name", "pattern": "  X  ",
                                              "unknown_key": 1, "timeout_sec": 3}})
    cfg = fa.load_config(fake)
    assert cfg["macro"] == "Python_XClick_Demo"           # healed to the default
    assert cfg["pattern"] == "X"
    assert "unknown_key" not in cfg
    assert cfg["timeout_sec"] == 15                       # 3 clamps into (15, 600)


def test_paths_info_follows_the_storage_mode(tmp_path):
    fake = make_bridge(tmp_path)
    from pathlib import Path
    cfg = dict(DEFAULTS)
    assert fa.paths_info(fake, cfg)["home"] == str(Path.home() / "Desktop" / "uivision")
    assert fa.paths_info(fake, cfg)["macro_file"].endswith("macros/Python_XClick_Demo.json")
    cfg["storage"] = "browser"
    assert fa.paths_info(fake, cfg)["macro_file"].endswith(
        str(fake.config.dir / "uivision" / "macros" / "Python_XClick_Demo.json"))


# ── run / stop slots ─────────────────────────────────────────────────────────

def test_run_slot_starts_once_and_refuses_the_second_run(tmp_path, monkeypatch):
    fake = make_bridge(tmp_path)
    scheduled = []
    monkeypatch.setattr(fa, "schedule_coro", lambda bridge, coro: scheduled.append(coro))
    res = json.loads(FirefoxAutoMixin.run_firefox_auto_test(fake))
    assert res == {"ok": True, "state": "running"}
    assert fake._firefox_auto_running is True and fake._firefox_auto_stop is False
    assert payloads(fake)[-1] == {"kind": "running"}
    second = json.loads(FirefoxAutoMixin.run_firefox_auto_test(fake))
    assert second["ok"] is False and "already in progress" in second["error"]
    assert len(scheduled) == 1
    scheduled[0].close()                                  # never awaited: close it


def test_stop_slot_needs_a_live_run(tmp_path):
    fake = make_bridge(tmp_path)
    res = json.loads(FirefoxAutoMixin.stop_firefox_auto_test(fake))
    assert res["ok"] is False and "no run in progress" in res["error"]
    fake._firefox_auto_running = True
    res = json.loads(FirefoxAutoMixin.stop_firefox_auto_test(fake))
    assert res == {"ok": True, "state": "stopping"}
    assert fake._firefox_auto_stop is True
    assert any(lvl == "warn" and "stop requested" in msg for lvl, msg in fake._logs)


async def test_do_run_test_streams_steps_and_the_verdict(tmp_path, monkeypatch):
    fake = make_bridge(tmp_path)
    fake._firefox_auto_running = True
    seen = {}

    async def fake_run(spec, report, seams=None):
        seen["spec"] = spec
        seen["seams"] = seams
        report("launch", "starting Firefox")
        return uiv_runner.RunResult(kind="ok", message="macro completed",
                                    steps=(("launch", "starting Firefox"),),
                                    lines=("echo: done",))

    monkeypatch.setattr(uiv_runner, "run_test", fake_run)
    await fa.do_run_test(fake)

    assert fake._firefox_auto_running is False            # finally clears the guard
    spec = seen["spec"]
    assert (spec.pattern, spec.macro, spec.storage) == ("Arena", "Python_XClick_Demo", "xfile")
    assert spec.config_dir == str(fake.config.dir)
    assert seen["seams"].stop() is False                  # stop flag wired, not pressed
    fake._firefox_auto_stop = True
    assert seen["seams"].stop() is True

    kinds = [p["kind"] for p in payloads(fake)]
    assert kinds == ["step", "step", "result"]            # run header + launch + verdict
    result = payloads(fake)[-1]
    assert result["result"] == "ok" and result["lines"] == ["echo: done"]
    assert result["steps"] == [["launch", "starting Firefox"]]
    assert any("🦊" in msg for _lvl, msg in fake._logs)   # RULE 2: every step logged


async def test_do_run_test_reports_an_explosion_as_blocked(tmp_path, monkeypatch):
    fake = make_bridge(tmp_path)
    fake._firefox_auto_running = True

    async def boom(spec, report, seams=None):
        raise RuntimeError("qt gone")

    monkeypatch.setattr(uiv_runner, "run_test", boom)
    await fa.do_run_test(fake)
    assert fake._firefox_auto_running is False
    result = payloads(fake)[-1]
    assert result["kind"] == "result" and result["result"] == "blocked"
    assert "qt gone" in result["message"]


def test_emit_status_survives_a_dead_signal(tmp_path):
    fake = make_bridge(tmp_path)

    def _boom(_payload):
        raise RuntimeError("qt gone")

    fake.firefox_auto_updated.emit = _boom
    fa.emit_status(fake, {"kind": "running"})             # best effort: no raise


def test_build_spec_maps_every_config_field(tmp_path):
    fake = make_bridge(tmp_path)
    cfg = dict(DEFAULTS, home="/h", binary="/b", timeout_sec=120, pause_ms=2000)
    spec = fa.build_spec(fake, cfg)
    assert spec == uiv_runner.RunSpec(
        pattern=cfg["pattern"], target=cfg["target"], macro=cfg["macro"],
        storage="xfile", home="/h", binary="/b", timeout_sec=120, pause_ms=2000,
        config_dir=str(fake.config.dir))
    assert not hasattr(spec, "url")                       # the spec no longer carries a URL
